### PR TITLE
Feature/217 project icon

### PR DESCRIPTION
**Pull request by niraymak**
## Description

Added:
 - FileController, FileService, FileRepository, FileModel
 - Extension: FileUploader which is responsible for handling uploads to the file server.
 - FileResourceResult, FileResource
 - AutoMapper for FileModel and FileResourceResult
 - Folder: files which is for storing files.
 - Dependency injection for FileRepository and FileService
 - Migrations for Database
- AllowedExtensionAttribute and MaxFileSizeAttribute

Updated:
 - .gitignore now ignores content of files folder
 - ProjectController now also accepts a reference to file (project-icon) on CreateProject and UpdateProject
 - Project Model (and ProjectResourceResult) now also has a file field (project-icon)

## Type of change

-   [ ] Bug fix (non-breaking change which fixes an issue)
-   [x] New feature (non-breaking change which adds functionality)
-   [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
-   [ ] This change requires a documentation update

## Checklist

-   [x] My code follows the style guidelines of this project
-   [x] I did not update API Controllers, if I did, I added/changed Postman tests
-   [x] I have performed a self-review of my own code
-   [x] I have commented my code, particularly in hard-to-understand areas
-   [x] I have made corresponding changes to the documentation
-   [x] I have added tests that prove my fix is effective or that my feature works
-   [x] New and existing unit tests pass locally with my changes
-   [x] I updated the changelog with an end-user readable description
-   [x] I assigned this pull request to the correct project board to update the sprint board

## Steps to Test or Reproduce

1. Run API and Identity server
2. Run runtests_windows.bat or runtests_linux.sh. (depending on your OS)
3. Check if all tests are running
4. Run UploadFile endpoint manually.
5. Check if file was uploaded to "API/wwwroot/Resources"

## Link to issue

Closes: #217

### CODE REVIEW

File: API/Controllers/FileController.cs
Comment (Maxvanhattum) [inline API/Controllers/FileController.cs:103]: I wouldn't straight take it from the request and use modelbinding, I would make a model and let asp.net check it
eg: 
`public async Task<IActionResult> UploadSingleFile([FromForm]File file)`
https://stackoverflow.com/questions/54424712/fromform-with-iformfile-using-asp-net-core

File: API/Controllers/ProjectController.cs
Comment (Maxvanhattum) [inline API/Controllers/ProjectController.cs:315]: Yet again I would use modelbinding

File: API/Extensions/FileUploader.cs
Comment (Maxvanhattum) [inline API/Extensions/FileUploader.cs:50]: Maybe check if path/file already exists and throw specific error if true

File: Postman/dex.postman_collection.json
Comment (Maxvanhattum) [inline Postman/dex.postman_collection.json:902]: Maybe set up one file included with the postman tests and configured in env variables so testing can always be done without manually having to configure the body with a file present on current tester host.

File: Postman/dex.postman_collection.json
Comment (Maxvanhattum) [inline Postman/dex.postman_collection.json:3111]: My postman tests don't pass: https://gyazo.com/8bec3b63972699b60b544264b2ad96a2 
I reckon this is because of above mentioned hardcoded file.

File: Services/Services/FileService.cs
Comment (Maxvanhattum) [inline Services/Services/FileService.cs:66]: Need to make the filename unique either here or in fileuploader but in my opinion it fits best in the service, I think this will prevent a lot of future headaches.

**Review by Maxvanhattum — CHANGES_REQUESTED** [review]
Few things that need to be changed. I think we should use modelbinding for getting the files from formdata, make the filename unique and check if by chance the path is already taken.
Also adding one 'test' file and configuring environment variable for the path.

File: API/Controllers/ProjectController.cs
Comment (niraymak) [reply API/Controllers/ProjectController.cs]: Changed it

File: API/Controllers/FileController.cs
Comment (niraymak) [reply API/Controllers/FileController.cs]: Changed it

File: Postman/dex.postman_collection.json
Comment (niraymak) [reply Postman/dex.postman_collection.json:902]: Added variables, unfortunately only works with CLI. Added .bat file and for now didn't yet make the right tests for file related requests. We should add them when postman tests are integrated with GitHub actions.

File: Postman/dex.postman_collection.json
Comment (niraymak) [reply Postman/dex.postman_collection.json]: Should pass now.

File: Services/Services/FileService.cs
Comment (niraymak) [reply Services/Services/FileService.cs:66]: Not yet done, should I change model and make path unique?

Comment (Maxvanhattum) [reply Services/Services/FileService.cs:66]: yea sounds good, on another note maybe some checking for illegal characters in the name and replace them with dashes. In another project we had to replace slashes etc since it screwed with the filesystem.

**Review by Maxvanhattum — CHANGES_REQUESTED** [review]
looks good, checking and making name unique still needs to be implemented. When it's done I'll run the postman tests again. Btw on the file&variable for postman i found a request for implementation but also a workaround: https://github.com/postmanlabs/postman-app-support/issues/3641#issuecomment-690281504

File: Services/Services/FileService.cs
Comment (niraymak) [reply Services/Services/FileService.cs:66]: > looks good, checking and making name unique still needs to be implemented. When it's done I'll run the postman tests again. Btw on the file&variable for postman i found a request for implementation but also a workaround: [postmanlabs/postman-app-support#3641 (comment)](https://github.com/postmanlabs/postman-app-support/issues/3641#issuecomment-690281504)

I used a very similar method. Unfortunately not working in Postman application for me.

Comment (Maxvanhattum) [reply Services/Services/FileService.cs:66]: Alright too bad, need to wait until it's implemented then

File: API/Controllers/FileController.cs
Comment (Maxvanhattum) [inline API/Controllers/FileController.cs:94]: This also requires to send all kinds of user data, these are required within the nested User object. If you go to swagger you see that those fields are required, while that info is gotten from the getcontextuser method on HttpContext.

File: API/Extensions/AllowedExtensionsAttribute.cs
Comment (Maxvanhattum) [inline API/Extensions/AllowedExtensionsAttribute.cs:11]: Missing XML summaries

File: API/Extensions/AllowedExtensionsAttribute.cs
Comment (Maxvanhattum) [inline API/Extensions/AllowedExtensionsAttribute.cs:23]: Move this line one down into the check if file is not null

File: API/Extensions/FileUploader.cs
Comment (Maxvanhattum) [inline API/Extensions/FileUploader.cs:125]: future iteration should check if the user making the request 'owns' the file to be deleted.

File: API/Extensions/MaxFileSizeAttribute.cs
Comment (Maxvanhattum) [inline API/Extensions/MaxFileSizeAttribute.cs:9]: needs XML summaries

File: Postman/dex.postman_collection.json
Comment (Maxvanhattum) [inline Postman/dex.postman_collection.json:1]: postman file tests dont work for me, I suggest making this another issue, since it's semi-complicated to fix and is holding up a working feature because of tech issues at the postman side.

File: Postman/runtests.bat
Comment (Maxvanhattum) [inline Postman/runtests.bat:1]: so this also doesnt work for me, the path is still local.

**Review by Maxvanhattum — CHANGES_REQUESTED** [review]
In my opinion the required user data fields need to be removed: https://gyazo.com/094882d35c19e7b20a5507d7d9d4fe8d
For deleting files it should check if the user 'owns' the file, or maybe if the user is an administrator?
When these are done I think this feature should be approved, removing the current postman tests, with the caveat that a separate issue for the postman tests is created.

File: API/Controllers/FileController.cs
Comment (niraymak) [reply API/Controllers/FileController.cs:94]: Maybe this happens because it asks for multipart-form-data? 

On delete is already check if the user owns the file or is administrator: https://gyazo.com/ecb4118f25c037fa58440b47d9f0baae
Although I see it should be " || " instead of "&& " right?

File: API/Controllers/ProjectController.cs
Comment (Brend-Smits) [inline API/Controllers/ProjectController.cs:111]: Fix formatting on this one please, TotalPages = await project......

File: API/Controllers/ProjectController.cs
Comment (Brend-Smits) [inline API/Controllers/ProjectController.cs:174]: Formatting is kinda messed up here as well, might be best to run a cleanup

File: API/Controllers/ProjectController.cs
Comment (Brend-Smits) [inline API/Controllers/ProjectController.cs:231]:
```suggestion
                file = await fileService.FindAsync(projectResource.FileId);
                project.FileId = file.Id
```

Comment (Brend-Smits) [reply API/Controllers/ProjectController.cs:231]: I think this is needed right? In case the file is updated, you want to update the file id of the project as well

File: API/Extensions/FileUploader.cs
Comment (Brend-Smits) [inline API/Extensions/FileUploader.cs:31]:
```suggestion
            : base($"File {name} already exists")
```

File: API/Extensions/FileUploader.cs
Comment (Brend-Smits) [inline API/Extensions/FileUploader.cs:81]:
```suggestion
            serverRoot = Path.Combine("https://localhost:5001/Resources/");
```
Never use hardcoded values like these. This should probably be a setting in appsettings (both dev and prod)

File: API/Extensions/FileUploader.cs
Comment (Brend-Smits) [inline API/Extensions/FileUploader.cs:138]:
```suggestion
        /// <returns> Bool which tells if file is deleted successfully or not </returns>
```

File: API/Extensions/MaxFileSizeAttribute.cs
Comment (Brend-Smits) [inline API/Extensions/MaxFileSizeAttribute.cs:55]:
```suggestion
            return $"Maximum allowed file size is {maxFileSize} bytes.";
```

File: API/Startup.cs
Comment (Brend-Smits) [inline API/Startup.cs:300]: Remove commented out code

File: API/Startup.cs
Comment (Brend-Smits) [inline API/Startup.cs:309]: Is this needed? If so, why?

File: CHANGELOG.md
Comment (Brend-Smits) [inline CHANGELOG.md:16]:
```suggestion
- Added a fileuploader which gives the opportunity to upload files and icons- [#217](https://github.com/DigitalExcellence/dex-backend/issues/217)
```

File: Models/File.cs
Comment (Brend-Smits) [inline Models/File.cs:31]: "this." is not needed.

File: Postman/local.postman_environment.json
Comment (Brend-Smits) [inline Postman/local.postman_environment.json:42]: I don't think these were meant to be changed. This goes for all changes in local.postman_environment.json
Please consider discarding changes in here.

File: Postman/runtests.bat
Comment (Brend-Smits) [inline Postman/runtests.bat:1]: This file needs to be converted to a bash file to work on linux.

File: Repositories/FileRepository.cs
Comment (Brend-Smits) [inline Repositories/FileRepository.cs:32]: Please comment everything, take the UserRepository as an example

File: Services/Services/FileService.cs
Comment (Brend-Smits) [inline Services/Services/FileService.cs:59]: This can be removed. It's no different than the inherited class.

File: Services/Services/FileService.cs
Comment (Brend-Smits) [inline Services/Services/FileService.cs:54]: This can be removed. It's no different than the inherited class.

File: Services/Services/FileService.cs
Comment (Brend-Smits) [inline Services/Services/FileService.cs:69]: Formatting

File: Services/Services/FileService.cs
Comment (Brend-Smits) [inline Services/Services/FileService.cs:31]: Please be sure to comment everything, take SourceManagerService.cs as an example

File: API/Controllers/FileController.cs
Comment (Brend-Smits) [inline API/Controllers/FileController.cs:182]: We should really use scopes instead of role checks like these. Keeps things organized. Please create a new scope and check for that. You can refer to UserController.cs#UpdateAccount as an example.

**Review by Brend-Smits — CHANGES_REQUESTED** [review]
I noticed there's an "APIFileFolderFile", what is this? (Check files changed).

I'm also missing Unit Tests for the FileService. Please add them. You can find them in 2_Services.Tests project. 

The rest looks solid, nice work!

File: .gitignore
Comment (niraymak) [reply .gitignore:338]: First line is the rule that ignores every file in the folder.
Second line makes sure not to ignore the .gitkeep file which makes sure that the folder is pushed.

File: API/Controllers/FileController.cs
Comment (niraymak) [reply API/Controllers/FileController.cs:94]: This is necessary because the request is form-data and not JSON.

File: Postman/dex.postman_collection.json
Comment (niraymak) [reply Postman/dex.postman_collection.json:1]: There was an issue with the variable. Discussed with Max in Slack and it works on his environment aswell nog.

File: API/1_API.csproj
Comment (niraymak) [reply API/1_API.csproj]: Restored

File: API/Controllers/ProjectController.cs
Comment (niraymak) [reply API/Controllers/ProjectController.cs:230]: It's automatically updated after saving the project.
Just checked it. Just updated id 1127: https://gyazo.com/50800f41fc637714c3a13e99003a378a

File: API/Startup.cs
Comment (niraymak) [reply API/Startup.cs:307]: I thought it was not working without it. Just tested and found out it is. So I removed it.

**Comment by Brend-Smits** [comment]
Since #240 is now merged, make sure the documentation is updated!

File: API/Extensions/FileUploader.cs
Comment (niraymak) [reply API/Extensions/FileUploader.cs:81]: removed lines as frontend is responsible for this.

File: API/Extensions/MaxFileSizeAttribute.cs
Comment (Maxvanhattum) [inline API/Extensions/MaxFileSizeAttribute.cs:11]: typo fort

File: API/Startup.cs
Comment (Maxvanhattum) [inline API/Startup.cs:292]: Ik weet niet wat hier precies in veranderd is/wat het doet

**Review by Maxvanhattum — APPROVED** [review]
Looks good to me, postman tests all work. Feature also works.

File: API/Startup.cs
Comment (niraymak) [reply API/Startup.cs:292]: Just compared with older version of this file. Only formatting is changed.

**Review by Brend-Smits — APPROVED** [review]
Tested, works well. Tests also look/work great. 
Thanks for the pr! 🎉

**Review by Brend-Smits — DISMISSED** [review]

File: API/Controllers/FileController.cs
Comment (Ruby77) [inline API/Controllers/FileController.cs:55]: `/// <param name="fileUploader">The file uploader extension</param>`

Remove the first `///`

File: API/Controllers/FileController.cs
Comment (Ruby77) [inline API/Controllers/FileController.cs:65]: I have been requested to create correct sentences for the summary. I think it's a good idea to continue this.

So change it to something like this:
`This method is responsible for retrieving a file`

File: API/Controllers/FileController.cs
Comment (Ruby77) [inline API/Controllers/FileController.cs:80]: Same applies here. Create a full correct sentence

File: API/Controllers/FileController.cs
Comment (Ruby77) [inline API/Controllers/FileController.cs:82]: In the returns tag you should show to return type in a success scenario. So in this example:
`This method returns status code 200` or something like this.

File: API/Extensions/AllowedExtensionsAttribute.cs
Comment (Ruby77) [inline API/Extensions/AllowedExtensionsAttribute.cs:71]: This is no problem at all!!! but maybe when you already are refactoring some things, you could remove this double enter.

**Review by Ruby77 — CHANGES_REQUESTED** [review]
Looks really solid! Good job!

I found a few really small improvement points (is not that much 😄). Overal note: for the documentation, try to create correct sentences with a dot at the end. Here a few examples I found:

**Comment by Brend-Smits** [comment]
Can we look into finishing this Pull Request so we can unblock the ones in the frontend as well? They are blocked because they need this merged first.

**Comment by Brend-Smits** [comment]
Sadly the Postman tests seem to be failing: https://github.com/DigitalExcellence/dex-backend/pull/241/checks?check_run_id=1332112727#step:6:8650

**Comment by niraymak** [comment]
Unfortunately this still does not work in production environment. I am not sure how to fix it. 

I reproduced this on my local machine (used docker environment), used postman to upload a file and it returns 200.
However, the delete file endpoint return "file does not exist". I tried to look into the files stored in docker environment but I couldn't find the file. 

So maybe something still goes wrong while uploading files in the docker enviroment.

**Review by Brend-Smits — APPROVED** [review]

**Review by Ruby77 — APPROVED** [review]
Awesome! 👍

**Review by Ruby77 — APPROVED** [review]

**Pull request merged**
